Add user defined metadata

criccomini opened the pull request:
# Which issue does this PR close?

Closes #4754

# Rationale for this change
 
We would like to be able to read and write user-defined metadata for AWS, GCP, and Azure. This use-case is mentioned briefly in #4754 before the conversation shifts to tags. It appears that tags were implemented, but user-defined metadata has not yet been.

# What changes are included in this PR?

- Add user-defined metadata for AWS, GCP, and Azure clients.

# Are there any user-facing changes?

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

I've added a `Metadata(String)` attribute to the `Attribute` enum. Other than that I don't believe there are any user-facing changes.

<!---
If there are any breaking changes to public APIs, please add the `breaking change` label.
-->

I do not believe there are any breaking changes. The user-facing change (`Attributes::Metadata(String)`) is additive and should be compatible. The one caveat here is I'm not sure how Rust handles enums (open vs. closed). If it handles them the way Java does (i.e. adding new enums can break old code) then this would be a breaking change. I'll leave it to the Rust experts to clarify this.

criccomini commented:
There are a few implementation details I'd like to call out:

1. The `reqwest` header name handling returns all custom headers as lower case. So if a user sets a `Metadata("FOO")`, the results returned in GetResult will have `Metadata("foo")`. This is somewhat unintuitive.
2. I elected to make `USER_DEFINED_METADATA_HEADER_PREFIX` a const rather than static because it's referenced in certain `const` spots. Rust didn't like `static` variables referenced from `const`.
3. I'm not thrilled about the `get.rs` code I added, but I couldn't figure out a better way. We need to iterate over all headers with the user-defined prefix (cloud-specific) and add them to `attributes`. I'm open to alternative approaches.
4. I did not add WebDav (http) support. The client simply ignores `Metadata` attributes on `put()` right now.

Xuanwo commented on object_store/src/http/client.rs:
```diff
@@ -183,6 +183,8 @@ impl Client {
                         builder.header(CONTENT_TYPE, v.as_ref())
                     }
+                    // Ignore metadata attributes
+                    Attribute::Metadata(_) => builder,
```
WebDAV handles custom metadata in `prop`, which differs from other object storage services. We can create an issue to track this.

tustvold replied:
In the meantime I think we should return an error as per https://docs.rs/object_store/latest/object_store/struct.PutOptions.html#structfield.attributes

Xuanwo commented on object_store/src/aws/client.rs:
```diff
@@ -642,6 +647,7 @@ impl GetClient for S3Client {
         etag_required: false,
         last_modified_required: false,
         version_header: Some(VERSION_HEADER),
+        user_defined_metadata_prefix: Some(USER_DEFINED_METADATA_HEADER_PREFIX),
```
nit: how about using `metadata_header_prefix`?

Xuanwo commented on object_store/src/aws/client.rs:
```diff
@@ -326,6 +327,10 @@ impl<'a> Request<'a> {
                     builder.header(CONTENT_TYPE, v.as_ref())
                 }
+                Attribute::Metadata(k_suffix) => builder.header(
+                    &format!("{}{}", USER_DEFINED_METADATA_HEADER_PREFIX, k_suffix),
```
Maybe we should use `self.config.metadata_header_prefix`?

criccomini replied:
`self.config` is `S3Config`, which does not have the metadata prefix. Are you suggesting we move `metadata_header_prefix` from HeaderConfig to S3Config (and same for GoogleCloudStorageConfig and so on)?

Xuanwo replied:
Oh, sorry for mixing up `HeaderCondig` and `S3Config`.

Xuanwo reviewed (commented):
This PR looks good to me overall!

Some missing works could be tracking in follow up:

- Add custom metadata support in multipart upload
- Add custom metadata integration test
- Track the current progress of webdav(http)
- Local fs might support custom metadata by `xattr`

Some design decision like naming need reviews from @tustvold and @alamb.

Xuanwo commented:
> The one caveat here is I'm not sure how Rust handles enums (open vs. closed).

Rust handles this by [non_exhaustive](https://doc.rust-lang.org/reference/attributes/type_system.html#the-non_exhaustive-attribute) attribute.

criccomini commented on object_store/src/aws/client.rs:
```diff
@@ -61,6 +61,7 @@ use std::sync::Arc;
 
 const VERSION_HEADER: &str = "x-amz-version-id";
 const SHA256_CHECKSUM: &str = "x-amz-checksum-sha256";
+const USER_DEFINED_METADATA_HEADER_PREFIX: &str = "x-amz-meta-";
```
Happy to rename to whatever. I chose to be quite verbose because I feel there's a lot of confusion about "metadata". You can see this play out in #4754, where the conversation muddles user-defined metadata and tags. Add in standard (non-user-defined) metadata and it gets even more murky.

Y'all let me know what to rename to, though. :)

criccomini commented:
Thanks @Xuanwo. I've updated the PR with your suggestions. I haven't renamed the variables yet since it sounds like @alamb or @tustvold need to weigh in.

Regarding your note about "Add custom metadata integration test", I was hoping [this change](https://github.com/apache/arrow-rs/pull/5915/files#diff-dceb79037f798cd3860cbfaa85c25b3fad940d2c743519f03ef10c2ca76969cdR483) would test the new functionality. Is it not enough?

alamb reviewed (commented):
THank you @criccomini  and @Xuanwo for the review

Please ping me if you need another review. I think @tustvold  may be back next week and he might get to reviewing this first

criccomini commented:
Sounds good! I'll check in next week. 😄

tustvold commented on object_store/src/attributes.rs:
```diff
@@ -45,6 +45,10 @@ pub enum Attribute {
+    /// Specifies a user-defined metadata field for the object
+    ///
+    /// The String is a user-defined key
+    Metadata(String),
```
It's likely a silliness that will make no measurable performance difference, but given these will often be static strings, I wonder what you think of making this `Cow<'static, str>`

tustvold commented on object_store/src/client/get.rs:
```diff
@@ -221,6 +221,20 @@ fn get_result<T: GetClient>(
+    if let Some(prefix) = T::HEADER_CONFIG.user_defined_metadata_prefix {
+        for (key, val) in response.headers() {
+            if let Some(suffix) = key.as_str().strip_prefix(prefix) {
+                if let Ok(val_str) = val.to_str() {
```
I wonder if we should error here?

tustvold reviewed (commented):
Just some minor nits, nice to see the benefits of all the time spent DRYing up the client logic :smile:

criccomini commented:
@tustvold I'm ready for another review here. I think I addressed all of your comments.

I had some issues with the `attributes.insert()` logic after I made the change. I ended up having to do `.to_string().into()` in order to disentangle the lifetimes (I think?). LMK if that's not the right approach here.

criccomini commented:
Gentle nudge here =)

tustvold reviewed (approved):
Looks good to me thank, sorry for the delay
